File: jas/tools/yaml_tool_selection_variants_test.py
"""Selection-VARIANT gesture-seam tests for the Python YAML tool runtime.

Ports the selection-variant seam tests from the Rust reference
(jas_dioxus/src/tools/yaml_tool.rs) for three selection tools that are
harder than the draw tools because they hit-test the document and (for
partial_selection) run an alt-drag-copy preview state machine:

  * partial_selection — control-point selection + marquee + the
    SEL-132 at-press / mid-drag alt-copy flow.
  * lasso — polygon enclosure selection + click-without-drag behavior.
  * interior_selection — click recurses into groups; marquee selects
    partially.

The base ``selection`` tool already works in this app, so this module
reuses the exact loader + model/context/hit-test/selection-seeding
machinery from ``yaml_tool_test.py``; only the cases are new. The
asserted child counts, selection paths, rect positions, and mode
strings mirror the Rust reference exactly.
"""

from __future__ import annotations

import json
import os
import sys

_REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
if _REPO_ROOT not in sys.path:
    sys.path.insert(0, _REPO_ROOT)
_JAS_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
if _JAS_DIR not in sys.path:
    sys.path.insert(0, _JAS_DIR)

import pytest

from document.controller import Controller
from document.document import Document, selection_kind_contains
from document.model import Model
from geometry.element import Group, Layer, Rect as RectElem


# ── Shared machinery (mirrors yaml_tool_test.py) ────────────────


def _load_ws_tool(tool_id: str) -> "YamlTool | None":
    from tools.yaml_tool import YamlTool

    ws_path = os.path.abspath(os.path.join(
        _REPO_ROOT, "workspace", "workspace.json",
    ))
    if not os.path.exists(ws_path):
        return None
    with open(ws_path, "r") as f:
        data = json.load(f)
    tools = data.get("tools")
    if not isinstance(tools, dict):
        return None
    spec = tools.get(tool_id)
    return YamlTool.from_workspace_tool(spec) if spec else None


def _ctx(model: Model):
    ctrl = Controller(model)
    ctx_obj = type("Ctx", (), {})()
    ctx_obj.model = model
    ctx_obj.controller = ctrl
    ctx_obj.document = model.document
    ctx_obj.request_update = lambda: None
    return ctx_obj, ctrl


def _children(model: Model):
    return model.document.layers[0].children


# ── Fixtures (mirror Rust fixtures) ─────────────────────────────


def _model_with_rect_element() -> Model:
    # Rect at (0, 0) 10x10 — control points:
    #   0 = (0, 0)   top-left
    #   1 = (10, 0)  top-right
    #   2 = (10, 10) bottom-right
    #   3 = (0, 10)  bottom-left
    layer = Layer(name="L", children=(
        RectElem(x=0.0, y=0.0, width=10.0, height=10.0),
    ))
    return Model(document=Document(layers=(layer,)))


def _selection_model_for_lasso() -> Model:
    # Single rect at (50, 50, 20, 20).
    layer = Layer(name="L", children=(
        RectElem(x=50.0, y=50.0, width=20.0, height=20.0),
    ))
    return Model(document=Document(layers=(layer,)))


def _model_with_rect_inside_group() -> Model:
    rect = RectElem(x=50.0, y=50.0, width=20.0, height=20.0)
    group = Group(children=(rect,))
    layer = Layer(name="L", children=(group,))
    return Model(document=Document(layers=(layer,)))


# ── Partial Selection variant tests ─────────────────────────────


class TestPartialSelectionVariants:
    def test_click_on_cp_selects_it(self):
        tool = _load_ws_tool("partial_selection")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _model_with_rect_element()
        ctx, _ = _ctx(model)
        # Click on CP 0 at (0, 0).
        tool.on_press(ctx, 0.0, 0.0)
        tool.on_release(ctx, 0.0, 0.0)
        sel = list(model.document.selection)
        assert len(sel) == 1
        assert sel[0].path == (0, 0)
        # The selection kind should include cp 0.
        assert selection_kind_contains(sel[0].kind, 0)
        assert model.can_undo

    def test_click_empty_starts_marquee(self):
        tool = _load_ws_tool("partial_selection")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _model_with_rect_element()
        ctx, _ = _ctx(model)
        # Click far from any CP.
        tool.on_press(ctx, 500.0, 500.0)
        # Mode should be "marquee".
        assert tool.tool_state("mode") == "marquee"
        # Release at a far position to commit the marquee.
        tool.on_release(ctx, 600.0, 600.0)
        # No hits → no selection.
        assert len(model.document.selection) == 0

    def test_marquee_picks_control_points(self):
        tool = _load_ws_tool("partial_selection")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _model_with_rect_element()
        ctx, _ = _ctx(model)
        # Marquee covering the rect's CPs (all at 0 or 10 in x and y).
        tool.on_press(ctx, -5.0, -5.0)
        tool.on_move(ctx, 15.0, 15.0, dragging=True)
        tool.on_release(ctx, 15.0, 15.0)
        # All 4 CPs of the rect should be selected (partial_select_rect
        # with extend=false replaces selection).
        sel = list(model.document.selection)
        assert len(sel) == 1
        assert sel[0].path == (0, 0)

    def test_at_press_alt_drag_copies_path(self):
        # SEL-132 at-press flow: with the rect selected, press on a CP
        # with Alt held, drag past threshold, release. Exactly one copy
        # is inserted (children 1 -> 2).
        tool = _load_ws_tool("partial_selection")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _model_with_rect_element()
        ctx, ctrl = _ctx(model)
        ctrl.select_element((0, 0))
        n_before = len(_children(model))
        tool.on_press(ctx, 0.0, 0.0, alt=True)
        tool.on_move(ctx, 5.0, 0.0, alt=True, dragging=True)
        tool.on_move(ctx, 80.0, 0.0, alt=True, dragging=True)
        tool.on_release(ctx, 80.0, 0.0, alt=True)
        n_after = len(_children(model))
        assert n_after == n_before + 1, (
            "alt-at-press drag should produce exactly one copy"
        )

    def test_mid_drag_alt_copies_path(self):
        # SEL-132 mid-drag flow: press WITHOUT Alt, drag past threshold,
        # press Alt mid-drag, release WITH Alt held. Same outcome as
        # at-press alt: exactly one copy inserted, original stays at its
        # original position (preview-restored on the alt-press
        # transition).
        tool = _load_ws_tool("partial_selection")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _model_with_rect_element()
        ctx, ctrl = _ctx(model)
        ctrl.select_element((0, 0))
        n_before = len(_children(model))
        tool.on_press(ctx, 0.0, 0.0)
        # Past 4-px threshold, no alt yet — snapshot fires, mode=moving,
        # translate by (5,0).
        tool.on_move(ctx, 5.0, 0.0, dragging=True)
        # Alt pressed mid-drag — entering preview: original snaps back
        # to (0,0,10,10).
        tool.on_move(ctx, 10.0, 0.0, alt=True, dragging=True)
        tool.on_move(ctx, 80.0, 0.0, alt=True, dragging=True)
        # Release with Alt still held — commit copy at cursor's release
        # position relative to the press position.
        tool.on_release(ctx, 80.0, 0.0, alt=True)
        children = _children(model)
        assert len(children) == n_before + 1, (
            "mid-drag alt + release-with-alt should commit exactly one copy"
        )
        # Original at (0,0) unchanged — preview snapped it back.
        original = children[0]
        assert isinstance(original, RectElem)
        assert original.x == 0.0, "original x preserved by preview restore"
        assert original.y == 0.0, "original y preserved by preview restore"
        # Copy at (80, 0) — translated by (cursor - press).
        copy = children[1]
        assert isinstance(copy, RectElem)
        assert copy.x == 80.0, "copy at cursor x"
        assert copy.y == 0.0, "copy at cursor y"

    def test_mid_drag_alt_preview_shows_real_copy(self):
        # During the mid-drag alt-preview phase, the document should
        # contain BOTH the original (snapped back to press) AND a real
        # copy at the cursor — so the user sees a moving rendered
        # element, not a bounding-box ghost.
        tool = _load_ws_tool("partial_selection")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _model_with_rect_element()
        ctx, ctrl = _ctx(model)
        ctrl.select_element((0, 0))
        n_before = len(_children(model))
        tool.on_press(ctx, 0.0, 0.0)
        tool.on_move(ctx, 5.0, 0.0, dragging=True)
        # Alt pressed mid-drag — enter preview.
        tool.on_move(ctx, 30.0, 0.0, alt=True, dragging=True)
        children = _children(model)
        assert len(children) == n_before + 1, (
            "during preview the document holds original + real copy"
        )
        original = children[0]
        assert isinstance(original, RectElem)
        assert original.x == 0.0, "original snapped back to press"
        copy = children[1]
        assert isinstance(copy, RectElem)
        assert copy.x == 30.0, "copy at cursor delta from press"

    def test_mid_drag_alt_released_before_mouseup_no_copy(self):
        # Press WITHOUT Alt, drag past threshold, press Alt mid-drag,
        # RELEASE Alt before mouseup. Should be a normal move — the
        # exit-preview transition re-applies the cumulative delta so the
        # original lands at the cursor; no copy is created.
        tool = _load_ws_tool("partial_selection")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _model_with_rect_element()
        ctx, ctrl = _ctx(model)
        ctrl.select_element((0, 0))
        n_before = len(_children(model))
        tool.on_press(ctx, 0.0, 0.0)
        tool.on_move(ctx, 5.0, 0.0, dragging=True)
        tool.on_move(ctx, 30.0, 0.0, alt=True, dragging=True)
        # Alt released before mouseup — exit preview: translate by
        # cumulative-from-press to land original at the cursor.
        tool.on_move(ctx, 50.0, 0.0, alt=False, dragging=True)
        tool.on_release(ctx, 50.0, 0.0, alt=False)
        children = _children(model)
        assert len(children) == n_before, (
            "alt-released-before-mouseup is a normal move; no copy"
        )
        original = children[0]
        assert isinstance(original, RectElem)
        assert original.x == 50.0, "original moved to cursor x"
        assert original.y == 0.0, "original y unchanged"


# ── Lasso variant tests ─────────────────────────────────────────


class TestLassoVariants:
    def test_lasso_select(self):
        tool = _load_ws_tool("lasso")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _selection_model_for_lasso()
        ctx, _ = _ctx(model)
        # Polygon enclosing the rect.
        tool.on_press(ctx, 40.0, 40.0)
        tool.on_move(ctx, 80.0, 40.0, dragging=True)
        tool.on_move(ctx, 80.0, 80.0, dragging=True)
        tool.on_move(ctx, 40.0, 80.0, dragging=True)
        tool.on_release(ctx, 40.0, 80.0)
        assert len(model.document.selection) != 0

    def test_lasso_miss(self):
        tool = _load_ws_tool("lasso")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _selection_model_for_lasso()
        ctx, _ = _ctx(model)
        # Polygon nowhere near the rect.
        tool.on_press(ctx, 0.0, 0.0)
        tool.on_move(ctx, 10.0, 0.0, dragging=True)
        tool.on_move(ctx, 10.0, 10.0, dragging=True)
        tool.on_move(ctx, 0.0, 10.0, dragging=True)
        tool.on_release(ctx, 0.0, 10.0)
        assert len(model.document.selection) == 0

    def test_click_without_drag_clears(self):
        tool = _load_ws_tool("lasso")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _selection_model_for_lasso()
        ctx, ctrl = _ctx(model)
        ctrl.select_element((0, 0))
        assert len(model.document.selection) != 0
        # Press + release at same point, no shift — buffer has 1 point,
        # fewer than 3 → falls into "clear selection" branch.
        tool.on_press(ctx, 5.0, 5.0)
        tool.on_release(ctx, 5.0, 5.0)
        assert len(model.document.selection) == 0

    def test_click_without_drag_shift_preserves(self):
        tool = _load_ws_tool("lasso")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _selection_model_for_lasso()
        ctx, ctrl = _ctx(model)
        ctrl.select_element((0, 0))
        # Shift+click without drag — the "clear selection" else-branch is
        # guarded by not shift_held so nothing happens.
        tool.on_press(ctx, 5.0, 5.0, shift=True)
        tool.on_release(ctx, 5.0, 5.0, shift=True)
        assert len(model.document.selection) != 0

    def test_state_transitions(self):
        tool = _load_ws_tool("lasso")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _selection_model_for_lasso()
        ctx, _ = _ctx(model)
        assert tool.tool_state("mode") == "idle"
        tool.on_press(ctx, 10.0, 10.0)
        assert tool.tool_state("mode") == "drawing"
        tool.on_release(ctx, 10.0, 10.0)
        assert tool.tool_state("mode") == "idle"


# ── Interior Selection variant tests ────────────────────────────


class TestInteriorSelectionVariants:
    def test_click_enters_group(self):
        tool = _load_ws_tool("interior_selection")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _model_with_rect_inside_group()
        ctx, _ = _ctx(model)
        # Click inside the rect (which lives at layer[0]/group[0]/rect[0]).
        tool.on_press(ctx, 55.0, 55.0)
        tool.on_release(ctx, 55.0, 55.0)
        sel = list(model.document.selection)
        assert len(sel) == 1
        assert sel[0].path == (0, 0, 0), (
            "interior selection should pick the leaf inside the group"
        )

    def test_marquee_selects_partial(self):
        tool = _load_ws_tool("interior_selection")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _model_with_rect_inside_group()
        ctx, _ = _ctx(model)
        tool.on_press(ctx, 40.0, 40.0)
        tool.on_move(ctx, 80.0, 80.0, dragging=True)
        tool.on_release(ctx, 80.0, 80.0)
        # partial_select_in_rect produced a selection; entries are
        # Partial so even whole-box coverage lists the element with
        # partial control points.
        assert len(model.document.selection) != 0

File: jas_ocaml/test/tools/yaml_tool_selection_variants_test.ml
(** Combined selection-VARIANT gesture-seam tests. Drives [Yaml_tool]
    with the partial_selection / lasso / interior_selection specs loaded
    from workspace/workspace.json and verifies behavior matches the
    Rust reference (jas_dioxus/src/tools/yaml_tool.rs). These are the
    harder selection variants: they hit-test the live document and (for
    partial_selection) run an alt-drag-copy preview state machine.

    Reuses the exact loader + model/ToolContext/hit-test setup from
    yaml_selection_tool_test.ml; only the cases are new. The probe
    builtins (doc.path.probe_partial_hit, hit_test_deep) read the
    registered document directly, so the always-miss hit-test stubs in
    [make_ctx] are inert for these tools. *)

open Jas

let () = ignore (GMain.init ())

(* ── Loaders ───────────────────────────────────────── *)

let load_tool (key : string) : Yaml_tool.yaml_tool option =
  match Workspace_loader.load () with
  | None -> None
  | Some ws ->
    match Workspace_loader.json_member "tools" ws.data with
    | Some (`Assoc tools) ->
      (match List.assoc_opt key tools with
       | Some spec -> Yaml_tool.from_workspace_tool spec
       | None -> None)
    | _ -> None

let partial_selection_tool () = load_tool "partial_selection"
let lasso_tool () = load_tool "lasso"
let interior_selection_tool () = load_tool "interior_selection"

(* ── Fixtures ──────────────────────────────────────── *)

let make_rect x y w h =
  Element.Rect { name = None; id = None;
    x; y; width = w; height = h; rx = 0.0; ry = 0.0;
    fill = None; stroke = None; opacity = 1.0;
    transform = None; locked = false; visibility = Preview;
    blend_mode = Normal; mask = None;
    fill_gradient = None; stroke_gradient = None;
  }

(** Single layer with one rect at (x,y,w,h). Mirrors Rust
    [model_with_rect_at]. *)
let model_with_rect_at x y w h : Model.model =
  let layer = Element.Layer {
    name = Some "L";
    id = None;
    children = [| make_rect x y w h |];
    transform = None; locked = false; opacity = 1.0;
    visibility = Preview; blend_mode = Normal; mask = None;
    isolated_blending = false; knockout_group = false;
  } in
  let m = Model.create () in
  m#set_document_unbracketed (Document.make_document [| layer |]);
  m

(** Rect at (0,0,10,10). Control points:
    0 = (0,0)  1 = (10,0)  2 = (10,10)  3 = (0,10). Mirrors Rust
    [model_with_rect_element]. *)
let model_with_rect_element () : Model.model =
  model_with_rect_at 0.0 0.0 10.0 10.0

(** Single rect at (50,50,20,20). Mirrors Rust
    [selection_parity_model_for_lasso]. *)
let model_for_lasso () : Model.model =
  model_with_rect_at 50.0 50.0 20.0 20.0

(** Layer -> Group -> Rect at (50,50,20,20); the rect lives at path
    [0;0;0]. Mirrors Rust [model_with_rect_inside_group]. *)
let model_with_rect_inside_group () : Model.model =
  let rect = make_rect 50.0 50.0 20.0 20.0 in
  let group = Element.Group {
    name = None; id = None;
    children = [| rect |];
    opacity = 1.0; transform = None; locked = false;
    visibility = Preview; blend_mode = Normal; mask = None;
    isolated_blending = false; knockout_group = false;
  } in
  let layer = Element.Layer {
    name = Some "L"; id = None;
    children = [| group |];
    transform = None; locked = false; opacity = 1.0;
    visibility = Preview; blend_mode = Normal; mask = None;
    isolated_blending = false; knockout_group = false;
  } in
  let m = Model.create () in
  m#set_document_unbracketed (Document.make_document [| layer |]);
  m

let make_ctx model =
  let ctrl = new Controller.controller ~model () in
  let ctx : Canvas_tool.tool_context = {
    model;
    controller = ctrl;
    hit_test_selection = (fun _ _ -> false);
    hit_test_handle = (fun _ _ -> None);
    hit_test_text = (fun _ _ -> None);
    hit_test_path_curve = (fun _ _ -> None);
    request_update = (fun () -> ());
    draw_element_overlay = (fun _cr _elem ~is_partial:_ _cps -> ());
  } in
  (ctx, ctrl)

(** First-layer children of the model document. *)
let layer0_children (m : Model.model) : Element.element array =
  match m#document.layers.(0) with
  | Element.Layer { children; _ } -> children
  | _ -> assert false

let rect_xy = function
  | Element.Rect { x; y; _ } -> (x, y)
  | _ -> assert false

(* ── Partial Selection ─────────────────────────────── *)

let partial_tests = [
  Alcotest.test_case "click_on_cp_selects_it" `Quick (fun () ->
    match partial_selection_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_with_rect_element () in
      let (ctx, _) = make_ctx m in
      (* Click on CP 0 at (0,0). *)
      tool#on_press ctx 0.0 0.0 ~shift:false ~alt:false;
      tool#on_release ctx 0.0 0.0 ~shift:false ~alt:false;
      let sel = m#document.selection in
      assert (Document.PathMap.cardinal sel = 1);
      (match Document.PathMap.find_opt [0; 0] sel with
       | Some es -> assert (Document.selection_kind_contains es.es_kind 0)
       | None -> assert false));

  Alcotest.test_case "click_empty_starts_marquee" `Quick (fun () ->
    match partial_selection_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_with_rect_element () in
      let (ctx, _) = make_ctx m in
      (* Click far from any CP -> marquee. *)
      tool#on_press ctx 500.0 500.0 ~shift:false ~alt:false;
      assert (tool#tool_state "mode" = `String "marquee");
      (* Release far away -> no hits -> empty selection. *)
      tool#on_release ctx 600.0 600.0 ~shift:false ~alt:false;
      assert (Document.PathMap.is_empty m#document.selection));

  Alcotest.test_case "marquee_picks_control_points" `Quick (fun () ->
    match partial_selection_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_with_rect_element () in
      let (ctx, _) = make_ctx m in
      (* Marquee covering all 4 CPs (at 0 or 10 in x and y). *)
      tool#on_press ctx (-. 5.0) (-. 5.0) ~shift:false ~alt:false;
      tool#on_move ctx 15.0 15.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_release ctx 15.0 15.0 ~shift:false ~alt:false;
      let sel = m#document.selection in
      assert (Document.PathMap.cardinal sel = 1);
      assert (Document.PathMap.mem [0; 0] sel));

  Alcotest.test_case "at_press_alt_drag_copies_path" `Quick (fun () ->
    match partial_selection_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_with_rect_element () in
      let (ctx, ctrl) = make_ctx m in
      (* Whole-element selection, then press on CP 0 with Alt, drag
         past threshold, release. Exactly one copy inserted. *)
      ctrl#select_element [0; 0];
      let n_before = Array.length (layer0_children m) in
      tool#on_press ctx 0.0 0.0 ~shift:false ~alt:true;
      tool#on_move ctx 5.0 0.0 ~shift:false ~alt:true ~dragging:true;
      tool#on_move ctx 80.0 0.0 ~shift:false ~alt:true ~dragging:true;
      tool#on_release ctx 80.0 0.0 ~shift:false ~alt:true;
      let n_after = Array.length (layer0_children m) in
      assert (n_after = n_before + 1));

  Alcotest.test_case "mid_drag_alt_copies_path" `Quick (fun () ->
    match partial_selection_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_with_rect_element () in
      let (ctx, ctrl) = make_ctx m in
      ctrl#select_element [0; 0];
      let n_before = Array.length (layer0_children m) in
      (* Press WITHOUT alt, drag past threshold (mode -> moving,
         translate by 5), press alt mid-drag (preview: original snaps
         back, real copy created), release WITH alt held. *)
      tool#on_press ctx 0.0 0.0 ~shift:false ~alt:false;
      tool#on_move ctx 5.0 0.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_move ctx 10.0 0.0 ~shift:false ~alt:true ~dragging:true;
      tool#on_move ctx 80.0 0.0 ~shift:false ~alt:true ~dragging:true;
      tool#on_release ctx 80.0 0.0 ~shift:false ~alt:true;
      let children = layer0_children m in
      assert (Array.length children = n_before + 1);
      (* Original at (0,0) — preview snapped it back. *)
      let (ox, oy) = rect_xy children.(0) in
      assert (ox = 0.0);
      assert (oy = 0.0);
      (* Copy at (80,0) — translated by (cursor - press). *)
      let (cx, cy) = rect_xy children.(1) in
      assert (cx = 80.0);
      assert (cy = 0.0));

  Alcotest.test_case "mid_drag_alt_preview_shows_real_copy" `Quick (fun () ->
    match partial_selection_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_with_rect_element () in
      let (ctx, ctrl) = make_ctx m in
      ctrl#select_element [0; 0];
      let n_before = Array.length (layer0_children m) in
      tool#on_press ctx 0.0 0.0 ~shift:false ~alt:false;
      tool#on_move ctx 5.0 0.0 ~shift:false ~alt:false ~dragging:true;
      (* Alt pressed mid-drag — enter preview; doc holds original +
         real copy. *)
      tool#on_move ctx 30.0 0.0 ~shift:false ~alt:true ~dragging:true;
      let children = layer0_children m in
      assert (Array.length children = n_before + 1);
      let (ox, _) = rect_xy children.(0) in
      assert (ox = 0.0);
      let (cx, _) = rect_xy children.(1) in
      assert (cx = 30.0));

  Alcotest.test_case "mid_drag_alt_released_before_mouseup_no_copy" `Quick (fun () ->
    match partial_selection_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_with_rect_element () in
      let (ctx, ctrl) = make_ctx m in
      ctrl#select_element [0; 0];
      let n_before = Array.length (layer0_children m) in
      tool#on_press ctx 0.0 0.0 ~shift:false ~alt:false;
      tool#on_move ctx 5.0 0.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_move ctx 30.0 0.0 ~shift:false ~alt:true ~dragging:true;
      (* Alt released before mouseup — exit preview; original lands at
         cursor; NO copy. *)
      tool#on_move ctx 50.0 0.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_release ctx 50.0 0.0 ~shift:false ~alt:false;
      let children = layer0_children m in
      assert (Array.length children = n_before);
      let (ox, oy) = rect_xy children.(0) in
      assert (ox = 50.0);
      assert (oy = 0.0));
]

(* ── Lasso ─────────────────────────────────────────── *)

let lasso_tests = [
  Alcotest.test_case "lasso_select" `Quick (fun () ->
    match lasso_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_for_lasso () in
      let (ctx, _) = make_ctx m in
      (* Polygon enclosing the rect at (50,50,20,20). *)
      tool#on_press ctx 40.0 40.0 ~shift:false ~alt:false;
      tool#on_move ctx 80.0 40.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_move ctx 80.0 80.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_move ctx 40.0 80.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_release ctx 40.0 80.0 ~shift:false ~alt:false;
      assert (not (Document.PathMap.is_empty m#document.selection)));

  Alcotest.test_case "lasso_miss" `Quick (fun () ->
    match lasso_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_for_lasso () in
      let (ctx, _) = make_ctx m in
      (* Polygon nowhere near the rect. *)
      tool#on_press ctx 0.0 0.0 ~shift:false ~alt:false;
      tool#on_move ctx 10.0 0.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_move ctx 10.0 10.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_move ctx 0.0 10.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_release ctx 0.0 10.0 ~shift:false ~alt:false;
      assert (Document.PathMap.is_empty m#document.selection));

  Alcotest.test_case "click_without_drag_clears" `Quick (fun () ->
    match lasso_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_for_lasso () in
      let (ctx, ctrl) = make_ctx m in
      ctrl#select_element [0; 0];
      assert (not (Document.PathMap.is_empty m#document.selection));
      (* Press + release at same point, no shift — buffer has 1 point,
         fewer than 3 -> clear-selection branch. *)
      tool#on_press ctx 5.0 5.0 ~shift:false ~alt:false;
      tool#on_release ctx 5.0 5.0 ~shift:false ~alt:false;
      assert (Document.PathMap.is_empty m#document.selection));

  Alcotest.test_case "click_without_drag_shift_preserves" `Quick (fun () ->
    match lasso_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_for_lasso () in
      let (ctx, ctrl) = make_ctx m in
      ctrl#select_element [0; 0];
      (* Shift+click without drag — shift_held captured at press; the
         clear-selection branch is guarded by not shift_held. *)
      tool#on_press ctx 5.0 5.0 ~shift:true ~alt:false;
      tool#on_release ctx 5.0 5.0 ~shift:true ~alt:false;
      assert (not (Document.PathMap.is_empty m#document.selection)));

  Alcotest.test_case "state_transitions" `Quick (fun () ->
    match lasso_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_for_lasso () in
      let (ctx, _) = make_ctx m in
      assert (tool#tool_state "mode" = `String "idle");
      tool#on_press ctx 10.0 10.0 ~shift:false ~alt:false;
      assert (tool#tool_state "mode" = `String "drawing");
      tool#on_release ctx 10.0 10.0 ~shift:false ~alt:false;
      assert (tool#tool_state "mode" = `String "idle"));
]

(* ── Interior Selection ────────────────────────────── *)

let interior_tests = [
  Alcotest.test_case "click_enters_group" `Quick (fun () ->
    match interior_selection_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_with_rect_inside_group () in
      let (ctx, _) = make_ctx m in
      (* Click inside the rect (layer[0]/group[0]/rect[0]). *)
      tool#on_press ctx 55.0 55.0 ~shift:false ~alt:false;
      tool#on_release ctx 55.0 55.0 ~shift:false ~alt:false;
      let sel = m#document.selection in
      assert (Document.PathMap.cardinal sel = 1);
      (* Interior selection picks the leaf inside the group, not the
         group itself. *)
      assert (Document.PathMap.mem [0; 0; 0] sel));

  Alcotest.test_case "marquee_selects_partial" `Quick (fun () ->
    match interior_selection_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = model_with_rect_inside_group () in
      let (ctx, _) = make_ctx m in
      tool#on_press ctx 40.0 40.0 ~shift:false ~alt:false;
      tool#on_move ctx 80.0 80.0 ~shift:false ~alt:false ~dragging:true;
      tool#on_release ctx 80.0 80.0 ~shift:false ~alt:false;
      assert (not (Document.PathMap.is_empty m#document.selection)));
]

let () =
  Alcotest.run "Yaml selection variants" [
    "Partial Selection", partial_tests;
    "Lasso", lasso_tests;
    "Interior Selection", interior_tests;
  ]
